### DISCUSSION TITLE
Resimulate with modified parameters via Scripting API

### DISCUSSION THREAD

**Discussion by hkiel** [OpenModelica/OpenModelica#11783] [Q&A] [answered]
I want to write an `oms` script to load a model and simulate it over and over with modified parameters.
I'd like to avoid recompiling the C code again and again and instead only modify some parameters and re-simulate, like it can be done manually in OMEdit.

What is the scripting API to modify parameters and resimulate/run the compiled model with these new parameters? I cannot find it in the documentation of the scripting API.

**Comment by casella** [comment]
@arun3688 this is similar to what you did with the OMPython linearize method for the MPC paper. Can you please comment on @hkiel's enquiry?

**Comment by arun3688** [comment]
@hkiel Yes you can do that, for example if you are using OMPython you can use the `ModelicaSystem` API which justs builds the model then you can use the `setParameter()` API to override the parameters multiple times and use the `simulate`  API to do the simulation. see the docs
https://openmodelica.org/doc/OpenModelicaUsersGuide/1.22/ompython.html#enhanced-ompython-features

An example script would be something like below
```
from OMPython import ModelicaSystem
mod = ModelicaSystem("C:/test/BouncingBall.mo", "BouncingBall")
mod.getParameters()
mod.setParameters("g=9.81")
mod.simulate()
```

**Reply by hkiel** [reply]
@arun3688 I'm not using OMPython, but I'm writing an oms script / working with OMNotebook. So, I'm using [OpenModelica.Scripting](https://build.openmodelica.org/Documentation/OpenModelica.Scripting.html).

**Comment by arun3688** [comment, marked as answer]
@hkiel then you should use buildModel() API, and then use the system() API to execute the simulation executable and override the variables
an example of mos script would

```
loadFile("BouncingBall.mo");
getErrorString();
buildModel(BouncingBall); // this will build the simulation exectuable
getErrorString();
system("./BouncingBall.exe -override=g=10") // this command will override the parameters 
```

**Reply by hkiel** [reply]
This is not too portable, as on Unix the executable will not be called `.exe`. However, I think it should be possible to skip the extension.
And I would have to create a string with all the parameters I want to modify. At least I can get a list of the model parameters with `getParameterNames()` and their values with `getParameterValue()`

**Reply by casella** [reply]
@hkiel, we could add a new API function that does what you need. Can you write down a specification for it?

**Reply by arun3688** [reply]
@hkiel if you are in linux then it should be without extension `system("./BouncingBall -override=g=10")`

**Reply by casella** [reply]
anyway, do we really need to call ./BouncingBall.exe? From the command line prompt, you just type the name of executables, without the .exe extension. I guess `system("./BouncingBall -override=g=10")` should work fine also on Windows.

**Reply by hkiel** [reply]
I just tried and the `"./"` at the beginning is making problems on Windows (`".\\"` works). You can leave out the `.exe` though.
`system("BouncingBall -override=g=10")` should do the trick. I'm not sure if that will work on Unixoids if `PATH` does not include `.` or another executable with the same name is on the path first.

**Comment by hkiel** [comment]
A specification could be like this function where parameterValue should be converted automatically from String to the required data type. If I recall correctly overloading functions is not supported in (Meta)Modelica!? It would be nice to pass the value in the correct form (data type) without the need to convert it to string before, so that you could pass e.g. an array of Integers if you need to.

```Modelica
function setParameterValue
  input TypeName class_;
  input String parameterName;
  input String parameterValue;
end setParameterValue;
```

**Reply by casella** [reply]
Could you also post an example of usage, please?

**Reply by arun3688** [reply]
@hkiel I guess we already have this API for `setParameters` but it is not exposed and also you can use that API only on model level, So you cannot override during runtime, to override during runtime you have to use `--override` flag
